### PR TITLE
[SPARK-33353][BUILD] Cache dependencies for Coursier with new sbt in GitHub Actions

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR change the behavior of GitHub Actions job that caches dependencies.
SPARK-33226 upgraded sbt to 1.4.1.
As of 1.3.0, sbt uses Coursier as the dependency resolver / fetcher.
So let's change the dependency cache configuration for the GitHub Actions job.

### Why are the changes needed?

To make build faster with Coursier for the GitHub Actions job.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Should be done by GitHub Actions itself.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Nice! I have been thinking about fixing it but I couldn't have a time to fix.

**Review by HyukjinKwon — APPROVED** [review]
Should be good to merge once we can verify they are cached. The changes look good to me.

Separately we should also cache `~/.sbt` too but it will be orthogonal.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -111,13 +111,13 @@ jobs:
         key: ${{ matrix.java }}-${{ matrix.hadoop }}-maven-${{ hashFiles('**/pom.xml') }}
         restore-keys: |
           ${{ matrix.java }}-${{ matrix.hadoop }}-maven-
-    - name: Cache Ivy local repository
+    - name: Cache Coursier local repository
       uses: actions/cache@v2
       with:
-        path: ~/.ivy2/cache
-        key: ${{ matrix.java }}-${{ matrix.hadoop }}-ivy-${{ hashFiles('**/pom.xml', '**/plugins.sbt') }}
+        path: ~/.cache/coursier
```
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:117]: Thank you, @sarutak . BTW, I cannot find this directory in my Mac. Is it dependent on OS?
```
$ ls ~/.cache/coursier
ls: /Users/dongjoon/.cache/coursier: No such file or directory
```

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:117]: The directories are different in OSs (https://get-coursier.io/docs/cache)

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:117]: Thanks! @HyukjinKwon .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (sarutak) [reply .github/workflows/build_and_test.yml:117]: > BTW, I cannot find this directory in my Mac. Is it dependent on OS?

Yeah, seems so.
On macOS, it is `~/Library/Caches/Coursier`.

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @sarutak and @HyukjinKwon !

**Pull request closed without merging**